airflow/models.py

```diff
@@ -4328,6 +4442,168 @@ def _test_cycle_helper(self, visit_map, task_id):
+                    TI.task_id.in_(self.task_ids),
+                    # Don't look for success/skip TIs. We check SLAs often, so
+                    # there's little chance that a TI switches to successful
+                    # after an SLA miss but before we notice; and this should
+                    # be a major perf boost (since most TIs are successful or
+                    # skipped).
+                    not_(TI.state.in_((State.SUCCESS, State.SKIPPED))),
+                    # Only look at specified DagRuns
+                    DR.id.in_(scheduled_dagrun_ids),
+                    # If the DAGRun is SUCCEEDED, then everything has gone
+                    # according to plan. But if it's FAILED, someone may be
+                    # coming to fix it, and SLAs for tasks in it will still
+                    # matter.
+                    DR.state != State.SUCCESS).all()
+        else:
+            scheduled_tis = []
+
+        # We need to examine unscheduled DAGRuns, too. If there are concurrency
+        # limitations, it's possible that a task instance will miss its SLA
+        # before its corresponding DAGRun even gets created.
+        last_dagrun = scheduled_dagruns[-1] if scheduled_dagruns else None
+
+        def unscheduled_tis(last_dagrun):
+            for dag_run in yield_unscheduled_runs(self, last_dagrun, ts):
+                for ti in yield_unscheduled_tis(dag_run, ts):
+                    yield ti
+
+        # Snapshot the time to check SLAs against.
+        ts = timezone.utcnow()
+
+        for ti in itertools.chain(scheduled_tis, unscheduled_tis(last_dagrun)):
+            ti.task = self.task_dict[ti.task_id]
+            # Ignore tasks that don't have SLAs, saving most calculation of
+            # future task instances.
+            if ti.task.has_slas():
+                create_sla_misses(ti, ts, session=session)
+
+        # Save any SlaMisses that were created in `create_sla_misses()`
+        session.commit()
+
+    @provide_session
+    def get_sla_notifications(self, session=None):
+        """
+        Find all SlaMisses for this DAG that haven't yet been notified.
+        """
+        return (
+            session
+            .query(SlaMiss)
+            .filter(SlaMiss.notification_sent == False)  # noqa
+            .filter(SlaMiss.dag_id == self.dag_id)
+            .all()
+        )
+
+    @provide_session
+    def send_sla_notifications(self, sla_misses, session=None):
+        """
+        Given a list of SLA misses, send emails and/or do SLA miss callback.
+        """
+        if not sla_misses:
+            self.log.info("send_sla_notifications was called without any SLA "
+                          "notifications to send!")
+            return
+
+        # Retrieve the context for this TI, but patch in the SLA miss object.
+        for sla_miss in sla_misses:
+            TI = TaskInstance
+            ti = session.query(TI).filter(
+                TI.dag_id != sla_miss.dag_id,
```

Comment:
In which case would TI.dag_id != sla_miss.dag_id ?

Reply:
Oops, that should also be `==` (so that we're querying for TIs matching the exact DAG ID, task ID, and execution date of the SLA miss).